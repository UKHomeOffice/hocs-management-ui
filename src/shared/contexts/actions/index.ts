import types from './types';

export interface Action<T> {
    type: string;
    payload: T;
}

export type ActionFunctionAny<R> = (...args: any[]) => R;

function createAction<Payload>(actionType: string) : ActionFunctionAny<Action<any>> {
    return (payload: Payload) => ({
        payload,
        type: actionType
    });
}

export const unsetError = createAction(types.UNSET_ERROR);

export const clearApiStatus = createAction(types.CLEAR_API_STATUS);
