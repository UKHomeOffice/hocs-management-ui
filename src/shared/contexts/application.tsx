import React, { Component } from 'react';
import types from './actions/types';
import { ContextAction } from './actions/index';
import { ErrorContent } from '../layouts/error';
import Config, { LayoutConfig } from '../models/config';
import ApiStatus from '../models/apiStatus';

export interface ApplicationState {
    apiStatus?: ApiStatus;
    csrf?: string;
    dispatch(action: ContextAction<any>): Promise<any>;
    error?: ErrorContent;
    layout?: LayoutConfig;
    hasRole(role: string): boolean;
    hasOneOfRoles(roles: string[]): boolean;
}

interface ApplicationProps {
    children: React.ReactElement;
    config: Config;
}

const defaultState: ApplicationState = {
    dispatch: () => Promise.resolve(),
    hasRole: (_: string) => { return false; },
    hasOneOfRoles: (_: string[]) => { return false; }
};

export const Context = React.createContext<ApplicationState>(defaultState);

const reducer = (state: ApplicationState, action: ContextAction<any>): ApplicationState => {
    let newState;
    switch (action.type) {

        case types.UNSET_ERROR:
            newState = { ...state, error: undefined };
            break;
        case types.CLEAR_API_STATUS:
            newState = { ...state, apiStatus: undefined };
            break;
        case types.UPDATE_API_STATUS:
            newState = { ...state, apiStatus: { ...action.payload } };
            break;
        default:
            console.warn('Unsupported action');
            newState = state;
    }
    return newState;
};

export class ApplicationProvider extends Component<ApplicationProps, ApplicationState> {

    constructor(props: ApplicationProps) {
        super(props);
        const { config } = props;

        console.log(props)

        this.state = {
            ...props.config,
            apiStatus: undefined,
            dispatch: (action: ContextAction<any>) => {
                try {
                    this.setState(state => reducer(state, action));
                    return Promise.resolve();
                } catch (error) {
                    return Promise.reject(error);
                }
            },
            hasRole: (role) => {
                return config.user?.roles.includes(role) || false;
            },
            hasOneOfRoles: (roles: string[]) => {
                return roles.some(role => config.user?.roles.includes(role)) || false;
            }
        };
    }

    render() {
        const { children } = this.props;
        return (
            <Context.Provider value={this.state}>
                {children}
            </Context.Provider>
        );
    }

}

export const ApplicationConsumer = Context.Consumer;
