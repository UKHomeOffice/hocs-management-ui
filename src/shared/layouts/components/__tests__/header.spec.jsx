import React from 'react';
import Header from '../header.tsx';
import { render } from '@testing-library/react';
import { ApplicationProvider } from '../../../contexts/application';
import { MemoryRouter } from 'react-router-dom';
import { HelmetProvider } from 'react-helmet-async';

describe('Layout header component', () => {
    it('should render with default props', () => {
        expect(
            render(
                <HelmetProvider>
                    <ApplicationProvider>
                        <MemoryRouter>
                            <Header />
                        </MemoryRouter>
                    </ApplicationProvider>
                </HelmetProvider>)
        ).toMatchSnapshot();
    });

    it('should render without crest when service is not GOV.UK', () => {
        expect(
            render(
                <HelmetProvider>
                    <ApplicationProvider>
                        <MemoryRouter>
                            <Header service="Test Service" />
                        </MemoryRouter>
                    </ApplicationProvider>
                </HelmetProvider>
            )
        ).toMatchSnapshot();
    });
});
