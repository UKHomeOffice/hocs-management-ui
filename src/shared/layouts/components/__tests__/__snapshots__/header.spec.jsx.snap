// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout header component should render with default props 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="govuk-header "
        data-module="header"
        role="banner"
      >
        <div
          class="govuk-header__container govuk-width-container"
        >
          <div
            class="govuk-header__container govuk-width-container"
          >
            <div
              class="govuk-header__logo"
            >
              <span
                class="govuk-header__logotype"
              >
                <a
                  class="govuk-header__link govuk-header__link--homepage govuk-header__logotype-text"
                  href="/"
                >
                  Correspondence Service
                </a>
              </span>
            </div>
            <div
              class="govuk-header__content"
            >
              <nav
                aria-label="Top Level Navigation"
                class="govuk-header__navigation "
              >
                <ul
                  class="govuk-header__navigation-list"
                  id="navigation"
                >
                  <li
                    class="govuk-header__navigation--end"
                  >
                    <a
                      class="govuk-header__link"
                      href="/oauth/logout"
                    >
                      Log out
                    </a>
                  </li>
                </ul>
              </nav>
            </div>
          </div>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="govuk-header "
      data-module="header"
      role="banner"
    >
      <div
        class="govuk-header__container govuk-width-container"
      >
        <div
          class="govuk-header__container govuk-width-container"
        >
          <div
            class="govuk-header__logo"
          >
            <span
              class="govuk-header__logotype"
            >
              <a
                class="govuk-header__link govuk-header__link--homepage govuk-header__logotype-text"
                href="/"
              >
                Correspondence Service
              </a>
            </span>
          </div>
          <div
            class="govuk-header__content"
          >
            <nav
              aria-label="Top Level Navigation"
              class="govuk-header__navigation "
            >
              <ul
                class="govuk-header__navigation-list"
                id="navigation"
              >
                <li
                  class="govuk-header__navigation--end"
                >
                  <a
                    class="govuk-header__link"
                    href="/oauth/logout"
                  >
                    Log out
                  </a>
                </li>
              </ul>
            </nav>
          </div>
        </div>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Layout header component should render without crest when service is not GOV.UK 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="govuk-header "
        data-module="header"
        role="banner"
      >
        <div
          class="govuk-header__container govuk-width-container"
        >
          <div
            class="govuk-header__container govuk-width-container"
          >
            <div
              class="govuk-header__logo"
            >
              <span
                class="govuk-header__logotype"
              >
                <a
                  class="govuk-header__link govuk-header__link--homepage govuk-header__logotype-text"
                  href="/"
                >
                  Test Service
                </a>
              </span>
            </div>
            <div
              class="govuk-header__content"
            >
              <nav
                aria-label="Top Level Navigation"
                class="govuk-header__navigation "
              >
                <ul
                  class="govuk-header__navigation-list"
                  id="navigation"
                >
                  <li
                    class="govuk-header__navigation--end"
                  >
                    <a
                      class="govuk-header__link"
                      href="/oauth/logout"
                    >
                      Log out
                    </a>
                  </li>
                </ul>
              </nav>
            </div>
          </div>
        </div>
      </header>
    </div>
  </body>,
  "container": <div>
    <header
      class="govuk-header "
      data-module="header"
      role="banner"
    >
      <div
        class="govuk-header__container govuk-width-container"
      >
        <div
          class="govuk-header__container govuk-width-container"
        >
          <div
            class="govuk-header__logo"
          >
            <span
              class="govuk-header__logotype"
            >
              <a
                class="govuk-header__link govuk-header__link--homepage govuk-header__logotype-text"
                href="/"
              >
                Test Service
              </a>
            </span>
          </div>
          <div
            class="govuk-header__content"
          >
            <nav
              aria-label="Top Level Navigation"
              class="govuk-header__navigation "
            >
              <ul
                class="govuk-header__navigation-list"
                id="navigation"
              >
                <li
                  class="govuk-header__navigation--end"
                >
                  <a
                    class="govuk-header__link"
                    href="/oauth/logout"
                  >
                    Log out
                  </a>
                </li>
              </ul>
            </nav>
          </div>
        </div>
      </div>
    </header>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
